pkg/volume/iscsi/iscsi_util.go

```diff
@@ -309,118 +323,145 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) (string, error) {
-			return "", err
-		}
 
-		if iscsiTransport == "" {
-			glog.Errorf("iscsi: could not find transport name in iface %s", b.Iface)
-			return "", fmt.Errorf("Could not parse iface file for %s", b.Iface)
+			if exist := waitForPathToExist(&devicePath, 10, iscsiTransport); !exist {
+				glog.Errorf("Could not attach disk: Timeout after 10s")
+				// update last error
+				lastErr = fmt.Errorf("Could not attach disk: Timeout after 10s")
+				continue
+			} else {
+				devicePaths[tp] = devicePath
+			}
 		}
-		if iscsiTransport == "tcp" {
-			devicePath = strings.Join([]string{"/dev/disk/by-path/ip", tp, "iscsi", b.Iqn, "lun", b.Lun}, "-")
-		} else {
-			devicePath = strings.Join([]string{"/dev/disk/by-path/pci", "*", "ip", tp, "iscsi", b.Iqn, "lun", b.Lun}, "-")
+		glog.V(4).Infof("iscsi: tried all devices for %q %d times, %d paths found", b.Iqn, i, len(devicePaths))
+		if len(devicePaths) == 0 {
+			// No path attached, report error and stop trying. kubelet will try again in a short while
+			// delete cloned iface
+			b.exec.Run("iscsiadm", "-m", "iface", "-I", b.Iface, "-o", "delete")
+			glog.Errorf("iscsi: failed to get any path for iscsi disk, last err seen:\n%v", lastErr)
+			return "", fmt.Errorf("failed to get any path for iscsi disk, last err seen:\n%v", lastErr)
+		}
+		if len(devicePaths) == len(bkpPortal) {
+			// We have all paths
+			glog.V(4).Infof("iscsi: all devices for %q found", b.Iqn)
+			break
 		}
-
-		if exist := waitForPathToExist(&devicePath, 10, iscsiTransport); !exist {
-			glog.Errorf("Could not attach disk: Timeout after 10s")
-			// update last error
-			lastErr = fmt.Errorf("Could not attach disk: Timeout after 10s")
-			continue
-		} else {
-			devicePaths = append(devicePaths, devicePath)
+		if len(devicePaths) >= minMultipathCount && i >= minAttachAttempts {
+			// We have at least two paths for multipath and we tried the other paths long enough
+			glog.V(4).Infof("%d devices found for %q", len(devicePaths), b.Iqn)
+			break
 		}
 	}
 
-	if len(devicePaths) == 0 {
-		// delete cloned iface
-		b.exec.Run("iscsiadm", "-m", "iface", "-I", b.Iface, "-o", "delete")
-		glog.Errorf("iscsi: failed to get any path for iscsi disk, last err seen:\n%v", lastErr)
-		return "", fmt.Errorf("failed to get any path for iscsi disk, last err seen:\n%v", lastErr)
-	}
 	if lastErr != nil {
 		glog.Errorf("iscsi: last error occurred during iscsi init:\n%v", lastErr)
 	}
 
+	devicePathList := []string{}
+	for _, path := range devicePaths {
+		devicePathList = append(devicePathList, path)
+	}
 	// Try to find a multipath device for the volume
-	if 1 < len(bkpPortal) {
-		// If the PV has 2 or more portals, wait up to 10 seconds for the multipath
-		// device to appear
-		devicePath = waitForMultiPathToExist(devicePaths, 10, b.deviceUtil)
+	if len(bkpPortal) > 1 {
+		// Multipath volume was requested. Wait up to 10 seconds for the multipath device to appear.
+		devicePath = waitForMultiPathToExist(devicePathList, 10, b.deviceUtil)
```

Comment:
Can we make `10` as a constant and avoid hardcoding?

Reply:
@jsafrane in case you missed above comment :)

Reply:
Changed to const.

Reply:
Thanks @jsafrane